qiskit/aqua/components/optimizers/aqgd.py

```diff
@@ -39,150 +41,307 @@ class AQGD(Optimizer):
+        return False
 
-        # store previous function evaluations
-        for i in range(n):
-            if i < n - 1:
-                self._previous_loss[i] = self._previous_loss[i + 1]
-            else:
-                self._previous_loss[i] = objval
+    def converged_alt(self, gradient: List[float], tol: float, window_size: int) -> bool:
+        """
+        Tests convergence from norm of windowed average of gradients
 
-        return False
+        Args:
+            gradient: current gradient
+            tol: tolerance for average gradient norm
+            window_size: size of averaging window
 
-    def optimize(self, num_vars, objective_function, gradient_function=None,
-                 variable_bounds=None, initial_point=None):
-        super().optimize(num_vars, objective_function, gradient_function,
-                         variable_bounds, initial_point)
+        Returns:
+            Bool indicating whether or not the optimization has converged
+        """
+        # If we haven't reached the required window length,
+        # append the current value, but we haven't converged
+        if len(self._prev_grad) < window_size - 1:
+            self._prev_grad.append(gradient)
+            return False
 
-        params = array(initial_point)
-        iter_count = 0
-        momentum = zeros(shape=(num_vars,))
-        objval = objective_function(params)
+        # Update last value in list with current value
+        self._prev_grad.append(gradient)
+        # (length now = n)
 
-        if self._disp:
-            print("Iteration: " + str(iter_count) + " \t| Energy: " + str(objval))
+        # Calculate previous windowed average
+        # and current windowed average of objective values
+        avg_grad = np.mean(self._prev_grad, axis=0)
 
-        minobj = objval
-        minparams = params
+        # Update window of values
+        # (Remove earliest value)
+        self._prev_grad.pop(0)
 
-        while iter_count < self._maxiter and not self.converged(objval):
-            for j in range(num_vars):
-                # update parameters in order based on quantum gradient
-                derivative = self.deriv(j, params, objective_function)
-                params, momentum[j] = self.update(j, params, derivative, momentum)
+        if np.linalg.norm(avg_grad, ord=np.inf) < tol:
+            # converged
+            logger.info("Avg. grad. norm: %f", np.linalg.norm(avg_grad, ord=np.inf))
+            return True
+        # else
+        return False
 
-            # check the value of the objective function
-            objval = objective_function(params)
+    def optimize(self,
+                 num_vars: int,
+                 objective_function: Callable,
+                 gradient_function: Callable = None,
+                 variable_bounds: List[Tuple[float, float]] = None,
+                 initial_point: np.ndarray = None) -> Tuple[np.ndarray, float, int]:
+        """
+        Perform optimization
```

Comment:
Again ideally the docstring should be inherited from the base class and not be replicated here unless there is some significant difference. Otherwise it has to be maintained in multiple places which is less than ideal. So I think this docstring can just be deleted from here and rely on the base class one.

Reply:
Done.